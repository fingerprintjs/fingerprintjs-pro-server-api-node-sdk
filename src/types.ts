import { components, paths } from './generatedApiTypes'

export enum Region {
  EU = 'EU',
  AP = 'AP',
  Global = 'Global',
}

export enum AuthenticationMode {
  AuthHeader = 'AuthHeader',
  QueryParameter = 'QueryParameter',
}

/**
 * Options for FingerprintJS server API client
 */
export interface Options {
  /**
   * Secret API key
   */
  apiKey: string
  /**
   * Region of the FingerprintJS service server
   */
  region?: Region
  /**
   * Authentication mode
   * Optional, default value is AuthHeader
   */
  authenticationMode?: AuthenticationMode

  /**
   * Optional fetch implementation
   * */
  fetch?: typeof fetch
}

/**
 * More info: https://dev.fingerprintjs.com/docs/server-api#query-parameters
 */
export type VisitorHistoryFilter = paths['/visitors/{visitor_id}']['get']['parameters']['query']

export type TooManyRequestsError = {
  status: 429
  /**
   * How many seconds to wait before retrying
   */
  retryAfter: number
}

/**
 * More info: https://dev.fingerprintjs.com/docs/server-api#response
 */
export type VisitorsResponse = paths['/visitors/{visitor_id}']['get']['responses']['200']['content']['application/json']
export type VisitorsResponse403 =
  paths['/visitors/{visitor_id}']['get']['responses']['403']['content']['application/json'] & {
    status: 403
  }
export type VisitorsResponse429 =
  paths['/visitors/{visitor_id}']['get']['responses']['429']['content']['application/json']

export type DeleteVisit404Response =
  paths['/visitors/{visitor_id}']['delete']['responses']['404']['content']['application/json']

export type DeleteVisit403Response =
  paths['/visitors/{visitor_id}']['delete']['responses']['403']['content']['application/json']

export type DeleteVisit400Response =
  paths['/visitors/{visitor_id}']['delete']['responses']['400']['content']['application/json'] & {
    status: 400
  }

export type CommonResponse429 = components['schemas']['ErrorCommon429Response']

export type VisitorsError = WithResponse<VisitorsResponse403 | VisitorsResponse429>

export type EventResponse = paths['/events/{request_id}']['get']['responses']['200']['content']['application/json']
export type EventResponse403 = paths['/events/{request_id}']['get']['responses']['403']['content']['application/json']
export type EventResponse404 = paths['/events/{request_id}']['get']['responses']['404']['content']['application/json']

type WithResponse<T> = T & {
  response: Response
}

/**
 * More info: https://dev.fingerprintjs.com/docs/webhooks#identification-webhook-object-format
 */
export type VisitWebhook = components['schemas']['WebhookVisit']
