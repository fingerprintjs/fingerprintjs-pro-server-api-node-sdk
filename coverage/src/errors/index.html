
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/errors</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> src/errors</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">91.01% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>81/89</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">63.63% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>35/55</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">76.47% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>26/34</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">93.97% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>78/83</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="apiErrors.ts"><a href="apiErrors.ts.html">apiErrors.ts</a></td>
	<td data-value="90.9" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 90%"></div><div class="cover-empty" style="width: 10%"></div></div>
	</td>
	<td data-value="90.9" class="pct high">90.9%</td>
	<td data-value="11" class="abs high">10/11</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="3" class="abs medium">2/3</td>
	<td data-value="90.9" class="pct high">90.9%</td>
	<td data-value="11" class="abs high">10/11</td>
	</tr>

<tr>
	<td class="file high" data-value="commonErrors.ts"><a href="commonErrors.ts.html">commonErrors.ts</a></td>
	<td data-value="87.5" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 87%"></div><div class="cover-empty" style="width: 13%"></div></div>
	</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="8" class="abs low">2/8</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	</tr>

<tr>
	<td class="file high" data-value="eventErrors.ts"><a href="eventErrors.ts.html">eventErrors.ts</a></td>
	<td data-value="90.47" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 90%"></div><div class="cover-empty" style="width: 10%"></div></div>
	</td>
	<td data-value="90.47" class="pct high">90.47%</td>
	<td data-value="21" class="abs high">19/21</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="24" class="abs medium">16/24</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="94.73" class="pct high">94.73%</td>
	<td data-value="19" class="abs high">18/19</td>
	</tr>

<tr>
	<td class="file high" data-value="getRetryAfter.ts"><a href="getRetryAfter.ts.html">getRetryAfter.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="unsealError.ts"><a href="unsealError.ts.html">unsealError.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="15" class="abs high">15/15</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="14" class="abs high">14/14</td>
	</tr>

<tr>
	<td class="file high" data-value="visitErrors.ts"><a href="visitErrors.ts.html">visitErrors.ts</a></td>
	<td data-value="87.09" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 87%"></div><div class="cover-empty" style="width: 13%"></div></div>
	</td>
	<td data-value="87.09" class="pct high">87.09%</td>
	<td data-value="31" class="abs high">27/31</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="18" class="abs medium">12/18</td>
	<td data-value="69.23" class="pct medium">69.23%</td>
	<td data-value="13" class="abs medium">9/13</td>
	<td data-value="92.85" class="pct high">92.85%</td>
	<td data-value="28" class="abs high">26/28</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2024-10-25T09:54:16.754Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    