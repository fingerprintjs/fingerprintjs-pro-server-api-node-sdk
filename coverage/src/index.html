
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> src</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">94.85% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>166/175</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">83.33% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>45/54</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">96.55% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>28/29</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">94.79% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>164/173</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	</tr>

<tr>
	<td class="file high" data-value="responseUtils.ts"><a href="responseUtils.ts.html">responseUtils.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	</tr>

<tr>
	<td class="file high" data-value="sealedResults.ts"><a href="sealedResults.ts.html">sealedResults.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="39" class="abs high">39/39</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="39" class="abs high">39/39</td>
	</tr>

<tr>
	<td class="file high" data-value="serverApiClient.ts"><a href="serverApiClient.ts.html">serverApiClient.ts</a></td>
	<td data-value="89.55" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 89%"></div><div class="cover-empty" style="width: 11%"></div></div>
	</td>
	<td data-value="89.55" class="pct high">89.55%</td>
	<td data-value="67" class="abs high">60/67</td>
	<td data-value="63.63" class="pct medium">63.63%</td>
	<td data-value="22" class="abs medium">14/22</td>
	<td data-value="90" class="pct high">90%</td>
	<td data-value="10" class="abs high">9/10</td>
	<td data-value="89.55" class="pct high">89.55%</td>
	<td data-value="67" class="abs high">60/67</td>
	</tr>

<tr>
	<td class="file high" data-value="types.ts"><a href="types.ts.html">types.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

<tr>
	<td class="file high" data-value="urlUtils.ts"><a href="urlUtils.ts.html">urlUtils.ts</a></td>
	<td data-value="96.96" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 96%"></div><div class="cover-empty" style="width: 4%"></div></div>
	</td>
	<td data-value="96.96" class="pct high">96.96%</td>
	<td data-value="33" class="abs high">32/33</td>
	<td data-value="91.66" class="pct high">91.66%</td>
	<td data-value="12" class="abs high">11/12</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="96.77" class="pct high">96.77%</td>
	<td data-value="31" class="abs high">30/31</td>
	</tr>

<tr>
	<td class="file medium" data-value="utils.ts"><a href="utils.ts.html">utils.ts</a></td>
	<td data-value="75" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 75%"></div><div class="cover-empty" style="width: 25%"></div></div>
	</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="4" class="abs medium">3/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="4" class="abs medium">3/4</td>
	</tr>

<tr>
	<td class="file high" data-value="webhook.ts"><a href="webhook.ts.html">webhook.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-06-04T13:30:03.627Z
            </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    