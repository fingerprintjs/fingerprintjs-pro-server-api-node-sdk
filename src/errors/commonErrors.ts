import { CommonResponse403, CommonResponse429 } from '../types'
import { getRetryAfter } from './getRetryAfter'
import { ApiError } from './apiErrors'

export class CommonError429 extends ApiError<429, CommonResponse429> {
  readonly retryAfter: number = 0

  constructor(body: CommonResponse429, response: Response) {
    super(body.error?.message ?? 'Too many requests', body, 429, body.error?.code ?? 'TooManyRequests', response)

    this.retryAfter = getRetryAfter(response)
  }
}

export class CommonError403 extends ApiError<403, CommonResponse403> {
  constructor(body: CommonResponse403, response: Response) {
    super(body.error?.message ?? 'Forbidden', body, 403, body.error?.code ?? 'Forbidden', response)
  }
}
