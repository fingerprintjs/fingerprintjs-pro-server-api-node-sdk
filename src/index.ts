export * from './urlUtils'
export * from './serverApiClient'
export * from './types'
export * from './sealedResults'
export * from './errors/unsealError'
export * from './webhook'
export * from './errors/apiErrors'
