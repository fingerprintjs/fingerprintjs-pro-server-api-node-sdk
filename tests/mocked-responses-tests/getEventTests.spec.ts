import { EventResponse403, EventResponse404, Region } from '../../src/types'
import { FingerprintJsServerApiClient } from '../../src/serverApiClient'
import getEventResponse from './mocked-responses-data/external/get_event_200.json'
import getEventWithExtraFieldsResponse from './mocked-responses-data/external/get_event_200_extra_fields.json'
import getEventAllErrorsResponse from './mocked-responses-data/external/get_event_200_all_errors.json'
import { EventError403, EventError404, SdkError } from '../../src/errors/apiErrors'

jest.spyOn(global, 'fetch')

const mockFetch = fetch as unknown as jest.Mock
describe('[Mocked response] Get Event', () => {
  const apiKey = 'dummy_api_key'
  const existingRequestId = '1626550679751.cVc5Pm'

  const client = new FingerprintJsServerApiClient({ region: Region.EU, apiKey })

  test('with request_id', async () => {
    mockFetch.mockReturnValue(Promise.resolve(new Response(JSON.stringify(getEventResponse))))

    const response = await client.getEvent(existingRequestId)

    expect(response).toMatchSnapshot()
  })

  test('with additional signals', async () => {
    mockFetch.mockReturnValue(Promise.resolve(new Response(JSON.stringify(getEventWithExtraFieldsResponse))))

    const response = await client.getEvent(existingRequestId)
    expect(response).toMatchSnapshot()
  })

  test('with all signals with failed error', async () => {
    mockFetch.mockReturnValue(Promise.resolve(new Response(JSON.stringify(getEventAllErrorsResponse))))

    const response = await client.getEvent(existingRequestId)

    expect(response).toMatchSnapshot()
  })

  test('403 error', async () => {
    const errorInfo = {
      error: {
        code: 'TokenRequired',
        message: 'secret key is required',
      },
    }
    const mockResponse = new Response(JSON.stringify(errorInfo), {
      status: 403,
    })
    mockFetch.mockReturnValue(Promise.resolve(mockResponse))
    await expect(client.getEvent(existingRequestId)).rejects.toThrow(
      new EventError403(errorInfo as EventResponse403, mockResponse)
    )
  })

  test('404 error', async () => {
    const errorInfo = {
      error: {
        code: 'RequestNotFound',
        message: 'request id is not found',
      },
    }
    const mockResponse = new Response(JSON.stringify(errorInfo), {
      status: 404,
    })
    mockFetch.mockReturnValue(Promise.resolve(mockResponse))
    await expect(client.getEvent(existingRequestId)).rejects.toThrow(
      new EventError404(errorInfo as EventResponse404, mockResponse)
    )
  })

  test('Error with bad shape', async () => {
    const errorInfo = 'Some text instead og shaped object'
    const mockResponse = new Response(
      JSON.stringify({
        error: errorInfo,
      }),
      {
        status: 404,
      }
    )
    mockFetch.mockReturnValue(Promise.resolve(mockResponse))
    await expect(client.getEvent(existingRequestId)).rejects.toThrow(EventError404)
    await expect(client.getEvent(existingRequestId)).rejects.toThrow('request id is not found')
  })

  test('Error with bad JSON', async () => {
    const mockResponse = new Response('(Some bad JSON)', {
      status: 404,
    })
    mockFetch.mockReturnValue(Promise.resolve(mockResponse))

    await expect(client.getEvent(existingRequestId)).rejects.toMatchObject(
      new SdkError(
        'Failed to parse JSON response',
        mockResponse,
        new SyntaxError('Unexpected token \'(\', \\"(Some bad JSON)\\" is not valid JSON')
      )
    )
  })
})
