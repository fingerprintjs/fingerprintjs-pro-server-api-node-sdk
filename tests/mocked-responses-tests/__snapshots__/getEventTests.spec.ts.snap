// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Mocked response] Get Event Error with bad shape 1`] = `
Object {
  "error": "{\\"error\\":\\"Some text instead og shaped object\\",\\"status\\":404}",
  "status": 0,
}
`;

exports[`[Mocked response] Get Event with additional signals 1`] = `
Object {
  "products": Object {
    "botd": Object {
      "data": Object {
        "bot": Object {
          "result": "notDetected",
        },
        "ip": "61.127.217.15",
        "time": "2019-05-21T16:40:13Z",
        "url": "https://www.example.com/login",
      },
    },
    "identification": Object {
      "data": Object {
        "browserDetails": Object {
          "browserFullVersion": "74.0.3729",
          "browserMajorVersion": "74",
          "browserName": "Chrome",
          "device": "Other",
          "os": "Windows",
          "osVersion": "7",
          "userAgent": "Mozilla/5.0 (Windows NT 6.1; Win64; x64) ....",
        },
        "confidence": Object {
          "score": 0.97,
        },
        "firstSeenAt": Object {
          "global": "2022-03-16T11:26:45.362Z",
          "subscription": "2022-03-16T11:31:01.101Z",
        },
        "incognito": true,
        "ip": "61.127.217.15",
        "ipLocation": Object {
          "accuracyRadius": 10,
          "city": Object {
            "name": "Dusseldorf",
          },
          "continent": Object {
            "code": "EU",
            "name": "Europe",
          },
          "country": Object {
            "code": "DE",
            "name": "Germany",
          },
          "latitude": 49.982,
          "longitude": 36.2566,
          "postalCode": "61202",
          "subdivisions": Array [
            Object {
              "isoCode": "63",
              "name": "North Rhine-Westphalia",
            },
          ],
          "timezone": "Europe/Dusseldorf",
        },
        "lastSeenAt": Object {
          "global": "2022-03-16T11:28:34.023Z",
          "subscription": null,
        },
        "linkedId": "somelinkedId",
        "requestId": "0KSh65EnVoB85JBmloQK",
        "time": "2019-05-21T16:40:13Z",
        "timestamp": 1582299576512,
        "url": "https://www.example.com/login",
        "visitorFound": true,
        "visitorId": "Ibk1527CUFmcnjLwIs4A9",
      },
    },
    "product3": Object {
      "data": Object {
        "result": false,
      },
    },
    "product4": Object {
      "data": Object {
        "details": Object {
          "detail1": true,
          "detail2": "detail description",
          "detail3": 42,
        },
        "result": true,
      },
    },
  },
}
`;

exports[`[Mocked response] Get Event with all signals with failed error 1`] = `
Object {
  "products": Object {
    "botd": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "clonedApp": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "emulator": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "factoryReset": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "frida": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "identification": Object {
      "error": Object {
        "code": "Failed",
        "message": "failed",
      },
    },
    "incognito": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "ipBlocklist": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "ipInfo": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "jailbroken": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "privacySettings": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "proxy": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "rawDeviceAttributes": Object {
      "data": Object {
        "audio": Object {
          "error": Object {
            "message": "internal server error",
            "name": "Error",
          },
        },
        "canvas": Object {
          "error": Object {
            "message": "internal server error",
            "name": "Error",
          },
        },
      },
    },
    "rootApps": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "tampering": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "tor": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "virtualMachine": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
    "vpn": Object {
      "error": Object {
        "code": "Failed",
        "message": "internal server error",
      },
    },
  },
}
`;

exports[`[Mocked response] Get Event with request_id 1`] = `
Object {
  "products": Object {
    "botd": Object {
      "data": Object {
        "bot": Object {
          "result": "notDetected",
        },
        "ip": "61.127.217.15",
        "time": "2019-05-21T16:40:13Z",
        "url": "https://www.example.com/login",
      },
    },
    "clonedApp": Object {
      "data": Object {
        "result": false,
      },
    },
    "emulator": Object {
      "data": Object {
        "result": false,
      },
    },
    "factoryReset": Object {
      "data": Object {
        "time": "1970-01-01T00:00:00Z",
        "timestamp": 0,
      },
    },
    "frida": Object {
      "data": Object {
        "result": false,
      },
    },
    "identification": Object {
      "data": Object {
        "browserDetails": Object {
          "browserFullVersion": "74.0.3729",
          "browserMajorVersion": "74",
          "browserName": "Chrome",
          "device": "Other",
          "os": "Windows",
          "osVersion": "7",
          "userAgent": "Mozilla/5.0 (Windows NT 6.1; Win64; x64) ....",
        },
        "confidence": Object {
          "score": 0.97,
        },
        "firstSeenAt": Object {
          "global": "2022-03-16T11:26:45.362Z",
          "subscription": "2022-03-16T11:31:01.101Z",
        },
        "incognito": true,
        "ip": "61.127.217.15",
        "ipLocation": Object {
          "accuracyRadius": 10,
          "city": Object {
            "name": "Dusseldorf",
          },
          "continent": Object {
            "code": "EU",
            "name": "Europe",
          },
          "country": Object {
            "code": "DE",
            "name": "Germany",
          },
          "latitude": 49.982,
          "longitude": 36.2566,
          "postalCode": "61202",
          "subdivisions": Array [
            Object {
              "isoCode": "63",
              "name": "North Rhine-Westphalia",
            },
          ],
          "timezone": "Europe/Dusseldorf",
        },
        "lastSeenAt": Object {
          "global": "2022-03-16T11:28:34.023Z",
          "subscription": null,
        },
        "linkedId": "somelinkedId",
        "requestId": "0KSh65EnVoB85JBmloQK",
        "time": "2019-05-21T16:40:13Z",
        "timestamp": 1582299576512,
        "url": "https://www.example.com/login",
        "visitorFound": true,
        "visitorId": "Ibk1527CUFmcnjLwIs4A9",
      },
    },
    "incognito": Object {
      "data": Object {
        "result": false,
      },
    },
    "ipBlocklist": Object {
      "data": Object {
        "details": Object {
          "attackSource": false,
          "emailSpam": false,
        },
        "result": false,
      },
    },
    "ipInfo": Object {
      "data": Object {
        "v4": Object {
          "address": "94.142.239.124",
          "asn": Object {
            "asn": "7922",
            "name": "COMCAST-7922",
            "network": "73.136.0.0/13",
          },
          "dataCenter": Object {
            "name": "DediPath",
            "result": true,
          },
          "geolocation": Object {
            "accuracyRadius": 20,
            "city": Object {
              "name": "Prague",
            },
            "continent": Object {
              "code": "EU",
              "name": "Europe",
            },
            "country": Object {
              "code": "CZ",
              "name": "Czechia",
            },
            "latitude": 50.05,
            "longitude": 14.4,
            "postalCode": "150 00",
            "subdivisions": Array [
              Object {
                "isoCode": "10",
                "name": "Hlavni mesto Praha",
              },
            ],
            "timezone": "Europe/Prague",
          },
        },
      },
    },
    "jailbroken": Object {
      "data": Object {
        "result": false,
      },
    },
    "privacySettings": Object {
      "data": Object {
        "result": false,
      },
    },
    "proxy": Object {
      "data": Object {
        "result": false,
      },
    },
    "rawDeviceAttributes": Object {
      "data": Object {
        "architecture": Object {
          "value": 127,
        },
        "audio": Object {
          "value": 35.73832903057337,
        },
        "canvas": Object {
          "value": Object {
            "Geometry": "4dce9d6017c3e0c052a77252f29f2b1c",
            "Text": "dd2474a56ff78c1de3e7a07070ba3b7d",
            "Winding": true,
          },
        },
        "colorDepth": Object {
          "value": 30,
        },
        "colorGamut": Object {
          "value": "srgb",
        },
        "contrast": Object {
          "value": 0,
        },
        "cookiesEnabled": Object {
          "value": true,
        },
      },
    },
    "rootApps": Object {
      "data": Object {
        "result": false,
      },
    },
    "tampering": Object {
      "data": Object {
        "anomalyScore": 0,
        "result": false,
      },
    },
    "tor": Object {
      "data": Object {
        "result": false,
      },
    },
    "virtualMachine": Object {
      "data": Object {
        "result": false,
      },
    },
    "vpn": Object {
      "data": Object {
        "methods": Object {
          "auxiliaryMobile": false,
          "publicVPN": false,
          "timezoneMismatch": false,
        },
        "result": false,
      },
    },
  },
}
`;
